### PR TITLE
Cipher refactory

### PR DESCRIPTION
**Pull request by sundapeng**
Refactory the concepts “codec” to "cipher+transformation"

### CODE REVIEW

File: src/main/java/com/intel/chimera/CryptoInputStream.java [hunk 1]
```diff
@@ -369,10 +369,25 @@ private void resetStreamOffset(long offset) throws IOException {
...
+  	int n = cipher.update(inBuffer, out);
+  	if (n < inputSize) {
+			/**
+			 * Typically code will not get here. Cipher#update will consume all 
+			 * input data and put result in outBuffer. 
+			 * Cipher#doFinal will reset the cipher context.
+			 */
+			cipher.doFinal(inBuffer, outBuffer);
+			cipherReset = true;
+		}
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/CryptoInputStream.java]: Indent 2 space for the if block above.

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ * Abstract Cipher class
+ */
+public abstract class Cipher {
+  public final static Logger LOG = LoggerFactory.getLogger(Cipher.class);
+
+  public static final int ENCRYPT_MODE = 1;
+  public static final int DECRYPT_MODE = 0;
+  
+  /**
+   * Get crypto codec for specified algorithm/mode/padding.
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: Get Cipher for ...

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+   * 
+   * @param props
+   *          the configuration properties
+   * @param cipherSuite
+   *          algorithm/mode/padding
+   * @return CryptoCodec the codec object. Null value will be returned if no
+   *         crypto codec classes with cipher suite configured.
+   */
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: codec to cipher

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        }
+      }
+    }
+    
+    if (cipher == null) {
+      // use JceCipher as the default
+      cipher = new JceCipher(props, transformation);
+    }
+
+    return cipher;
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: return (cipher==null)?new JceCipher(props, transformation):cipher;

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  /**
+   * Get crypto codec for algorithm/mode/padding in config value
+   * chimera.crypto.cipher.suite
+   * 
+   * @return CryptoCodec the codec object Null value will be returned if no
+   *         crypto codec classes with cipher suite configured.
+   */
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: cipher

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      cipher = new JceCipher(props, transformation);
+    }
+
+    return cipher;
+  }
+
+  /**
+   * Get crypto codec for algorithm/mode/padding in config value
+   * chimera.crypto.cipher.suite
+   *
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: Get cipher codec

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+   *         crypto codec classes with cipher suite configured.
+   */
+  public static Cipher getInstance() throws GeneralSecurityException {
+    return getInstance(new Properties());
+  }
+
+  /**
+   * Get crypto codec for algorithm/mode/padding in config value
+   * chimera.crypto.cipher.suite
+   *
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: Get cipher for ...

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+          "No cipher classes with cipher suite configured.");
+      return null;
+    }
+    for (String c : Splitter.on(',').trimResults().omitEmptyStrings().
+        split(cipherClassString)) {
+      try {
+        Class<?> cls = ReflectionUtils.getClassByName(c);
+        result.add(cls.asSubclass(Cipher.class));
+      } catch (ClassCastException e) {
+        LOG.debug("Class {} is not a Cipher.", c);
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: Log.error

File: src/main/java/com/intel/chimera/crypto/Cipher.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+      } catch (ClassNotFoundException e) {
+        LOG.debug("Cipher {} not found.", c);
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/Cipher.java]: Log.error

File: src/main/java/com/intel/chimera/crypto/JceCipher.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import com.google.common.base.Preconditions;
+import com.intel.chimera.utils.Utils;
+
+/**
+ * Implement the Cipher using JCE provider.
+ */
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/JceCipher.java]: Indent 2 space for this class. I will stop pointing this out.

File: src/main/java/com/intel/chimera/crypto/JceCipher.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+		this.provider = Utils.getJCEProvider(props);
+		this.transformation = transformation;
+		
+		final String secureRandomAlg = Utils.getSecureRandomAlg(props);
+		try {
+			random = (provider != null) ? 
+					SecureRandom.getInstance(secureRandomAlg, provider) : 
+						SecureRandom.getInstance(secureRandomAlg);
+		} catch (GeneralSecurityException e) {
+			LOG.warn(e.getMessage());
```
Comment (winningsix) [inline src/main/java/com/intel/chimera/crypto/JceCipher.java]: error

**Comment by sundapeng** [comment]
Hi @winningsix , thank you for your comments, I'm still fixing the typos in my local env, could you add a tag [typo] before your comment for these kind comments?

**Comment by winningsix** [comment]
@sundapeng OK. I see some code violates the code style like indents. Could you reformat them for the next updates? Thank you!

**Comment by jerrychenhf** [comment]
Originally, I thought to address the comments and styles incrementally. I think it is also a good idea to address the comments and format in the first merge.

**Comment by sundapeng** [comment]
Thank @winningsix for pointing it out, I will also check it before next update.

**Comment by sundapeng** [comment]
@jerrychenhf Okay, I will fix the comments and code styles on next update.

**Comment by winningsix** [comment]
@jerrychenhf @sundapeng The ReadMe file also needs to be updated. We can do it in a separate PR.

**Comment by sundapeng** [comment]
it's great, thank @winningsix

**Comment by sundapeng** [comment]
Patch updated with fixed comments, variables and style, I will merge the pr to master if there is no more question about the refactory. Thanks @jerrychenhf @winningsix @dianfu .

**Comment by winningsix** [comment]
Thanks @sundapeng LGTM

**Comment by jerrychenhf** [comment]
Thanks @sundapeng @winningsix 
Please add the java doc comment for the new abstract functions added in the Cipher class. You can got the comment from Java method.
After this, please merge. And we can have code check after you the merge.

**Comment by sundapeng** [comment]
I have added the javadoc for cipher, thank @jerrychenhf

**Comment by jerrychenhf** [comment]
It looks good. Thanks @sundapeng. Please go forward to merge.

**Comment by sundapeng** [comment]
PR Merged, thank @jerrychenhf @winningsix and @dianfu

**Pull request merged**